src/mapgen/mapgen.cpp

```diff
@@ -620,11 +623,22 @@ MapgenBasic::MapgenBasic(int mapgenid, MapgenParams *params, EmergeParams *emerg
 }
 
 
+MapgenBasic::MapgenBasic(const NodeDefManager *ndef_p) : Mapgen() {
+	ndef = ndef_p;
+
+	c_stone              = ndef_p->getId("mapgen_stone");
+	c_water_source       = ndef_p->getId("mapgen_water_source");
+	c_river_water_source = ndef_p->getId("mapgen_river_water_source");
```

Comment:
What about the other class members? Those will contain uninitialized data.

Reply:
They will, but I'm only setting what is actually used by `generateBiomes` (it uses a temporary object).

I set some class members from `l_generate_biomes()` but I need this function for the protected ones.

I can make `ModApiMapgen` a friend class of `MapgenBasic`, leave the definition empty and initialize all class members in `l_generate_biomes` if that's better.

Reply:
I haven't looked at the code changes yet but _in general_ it is better to refactor the code so that such workarounds are not even necessary.

Reply:
is it better with the new commit?

Reply:
(not yet; continued in another comment)